pkg/storage/ingest/reader.go

```diff
@@ -0,0 +1,331 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ingest
+
+import (
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/dskit/multierror"
+	"github.com/grafana/dskit/services"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/twmb/franz-go/pkg/kadm"
+	"github.com/twmb/franz-go/pkg/kerr"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/plugin/kprom"
+)
+
+// consumerGroup is only used to store commit offsets, not for actual consuming.
+const consumerGroup = "mimir"
+
+type record struct {
+	tenantID string
+	content  []byte
+}
+
+type recordConsumer interface {
+	consume(context.Context, []record) error
+}
+
+type PartitionReader struct {
+	services.Service
+
+	kafkaAddress  string
+	kafkaTopic    string
+	kafkaClientID string
+	partitionID   int32
+
+	client    *kgo.Client
+	admClient *kadm.Client
+
+	consumer recordConsumer
+	metrics  readerMetrics
+
+	commitInterval time.Duration
+	commitLoopWg   *sync.WaitGroup
+	commitFetches  chan kgo.Fetches
+
+	logger log.Logger
+}
+
+func NewReaderForPusher(kafkaAddress, kafkaTopic, kafkaClientID string, partitionID int32, pusher Pusher, logger log.Logger, reg prometheus.Registerer) (*PartitionReader, error) {
```

Comment:
[nit] Given it creates a PartitionReader, have you considered renaming this function and `newReader` to `NewPartitionReader...`?